components/script/dom/intersectionobserver.rs

```diff
@@ -164,42 +191,65 @@ impl IntersectionObserverMethods<crate::DomTypeHolder> for IntersectionObserver
         can_gc: CanGc,
         callback: Rc<IntersectionObserverCallback>,
         init: &IntersectionObserverInit,
-    ) -> DomRoot<IntersectionObserver> {
-        Self::new(window, proto, callback, init, can_gc)
-    }
-}
+    ) -> Fallible<DomRoot<IntersectionObserver>> {
+        // > 3. Attempt to parse a margin from options.rootMargin. If a list is returned,
+        // >    set this’s internal [[rootMargin]] slot to that. Otherwise, throw a SyntaxError exception.
+        let root_margin = if let Ok(margin) = parse_a_margin(init.rootMargin.as_ref()) {
+            margin
+        } else {
+            return Err(Error::Syntax)
+        };
 
-#[derive(JSTraceable, MallocSizeOf)]
-struct MarginRect {
-    top: LengthOrPercentage,
-    right: LengthOrPercentage,
-    bottom: LengthOrPercentage,
-    left: LengthOrPercentage,
-}
+        // > 4. Attempt to parse a margin from options.scrollMargin. If a list is returned,
+        // >    set this’s internal [[scrollMargin]] slot to that. Otherwise, throw a SyntaxError exception.
+        let scroll_margin = if let Ok(margin) = parse_a_margin(init.scrollMargin.as_ref()) {
+            margin
+        } else {
+            return Err(Error::Syntax)
+        };
 
-impl MarginRect {
-    fn zero() -> Self {
-        Self {
-            top: LengthOrPercentage::zero(),
-            right: LengthOrPercentage::zero(),
-            bottom: LengthOrPercentage::zero(),
-            left: LengthOrPercentage::zero(),
+        // > 5. Let thresholds be a list equal to options.threshold.
+        //
+        // Non-sequence value should be converted into Vec.
+        // Empty vec also generated for default value of threshold.
+        let mut threshold = match &init.threshold {
+            Some(DoubleOrDoubleSequence::Double(num)) => vec![*num],
+            Some(DoubleOrDoubleSequence::DoubleSequence(sequence)) => sequence.clone(),
+            None => Default::default(),
+        };
+
+        // > 6. If any value in thresholds is less than 0.0 or greater than 1.0, throw a RangeError exception.
+        threshold.iter().try_for_each(|num| {
```

Comment:
I think this would be easier to read as a for loop:
```rust
for num in &threshold {
    if **num < 0.0 || **num > 1.0 {
        return Err(Error::Range("Value in thresholds should not be less than 0.0 or greater than 1.0".to_owned()));
    }
}
```